Allows definition of each manufacturer's quirks

nao-pon opened the pull request:
It fixes #200.

SudarshanKulakrni commented:
Thank you so much @nao-pon san for your prompt help ! Highly appreciated. 

I just checked the edge version, the logs indicate following error.

```
This error originated from a custom integration.

Logger: custom_components.echonetlite
Source: custom_components/echonetlite/__init__.py:100
integration: ECHONETLite (documentation, issues)
First occurred: 10:32:29 PM (52 occurrences)
Last logged: 10:37:16 PM

[0x2(2)-0xa5(165)-0xf5(245)] - Unable to resolve the item name. Please report the unknown code [0x2(2)-0xa5(165)-0xf5(245)] at the issue tracker on GitHub!
[0x2(2)-0xa5(165)-0xf6(246)] - Unable to resolve the item name. Please report the unknown code [0x2(2)-0xa5(165)-0xf6(246)] at the issue tracker on GitHub!
```

nao-pon commented:
@SudarshanKulakrni Thanks for trying edge!

Are your additional entities configured correctly in HA?

The error may be displayed incorrectly. If the entity names are displayed correctly, that's fine.

SudarshanKulakrni commented:
Hi @nao-pon  Thanks for the quick trial again. 

I had installed the edge version in morning and waited for sunny weather to test entirely.
The new entities have started showing values, however I think these are too high. 
When I took attached screenshot, battery was around 95%, appliances were consuming 0.8 Kw (Solar panel providing 0.4 Kw and Battery providing 0.4), no consumption from grid.

![image](https://github.com/user-attachments/assets/284c64be-5078-4c3c-a72a-497482802712)

nao-pon commented:
@SudarshanKulakrni I made a mistake by confusing the hex string length with the binary data string length. I fixed that by updating edge to 0b28cf7. I also changed the unit of cumulative power to kWh.

Please download and try again. 😄

SudarshanKulakrni commented:
@nao-pon san,

Thats great ! The values are shown in Watts and I can see the values as in attached screenshot.
Now its evening, so no solar electricity production. I will verify correctness of the numbers when solar production is happening and system is charging/discharging battery. Lets see if we can see total of all inputs in F5. 
I will be able to test in daytime on Friday. Please allow me to report the results later.

**It is tough to fix bugs without having a debugging environment. Impressed with your skill :)** 

With this new edge version, there is a slight issue. When we start autodiscovery of devices, or manually provide IP address, it used to ask location of device (Living room, Bed room etc). Now it is not asking. This is a minor issue as we can add the device to areas later. 

Thanks again ! 

![image](https://github.com/user-attachments/assets/c324dc4b-dcec-4e40-b982-a51df4bb6c92)

nao-pon commented:
For now, I'm relieved that it seems to be working properly. The issue of not being able to set the device location during initial configuration was not reproducible in my environment, so if it was previously configured in HA, the previous saved setting value may have been referenced. Since that's a function of HA, I don't think it's an integration issue.

I'm grad to hear the results of additional testing at a later date. 👍

SudarshanKulakrni commented:
Hi @nao-pon I was able to grab the screenshots of the current status. The figures are not adding up.
None of the entities from PCS addup to 2.4kw as displayed on panel 😓

![image](https://github.com/user-attachments/assets/543126dd-4255-4b27-9c3c-3c36a2585fad)
![IMG_4105](https://github.com/user-attachments/assets/49ef9196-855d-4aaf-a309-8868717caefe)

nao-pon commented:
Turn on "Enable debug logging" on this integration screen in settings, and once you have acquired the data, turn off "Enable debug logging" to download the log. Please extract any parts of the log that you find useful, or attach the file as is. I would like to see what kind of raw data is being sent.

Also, calculations from the raw data are performed in the following file. It might be a good idea to tinker with that file.

- custom_components/echonetlite/quirks/Nichicon/all/02A5.py
```python
def _02A5F5(edt):
    d1 = d2 = None
    try:
        d1 = _signed_int(edt[0:4])
        d2 = _signed_int(edt[4:8])
    except:
        pass
    finally:
        return {"household_consumption": d1, "from/to_grid": d2}




def _02A5F6(edt):
    d1 = d2 = None
    try:
        d1 = float(_int(edt[0:4])) / 1000
        d2 = float(_int(edt[4:8])) / 1000
    except:
        pass
    finally:
        return {"normal_direction": d1, "reverse_direction": d2}
```

SudarshanKulakrni commented:
Hi @nao-pon  san,

Good morning. Thanks for your efforts to help the Nichicon customers. Today weather is nice and solar production is good. 
I disabled all other Echonet devices on my network to get a clear log of Nichicon, attaching the raw log. 
Also attaching 2 photos of solar panel's current stage, one with single input and one with all inputs added.

For capturing logs, after enabling logging, I reloaded the device to capture all refreshed values.
Please see if this helps. I will try to capture the log at later part of the day to have variation in test data.
 
![9209a2aa-f999-4744-92ed-eea7969a135a](https://github.com/user-attachments/assets/363976ae-d8e4-42fd-a89e-6acf605c89c7)
![3f286d84-a43b-481e-89f5-cab1cfacb2b5](https://github.com/user-attachments/assets/2744ce90-31c1-47be-bb8c-563adf97f6fb)
[home-assistant_echonetlite_2025-02-08T01-47-58.105Z.log](https://github.com/user-attachments/files/18716437/home-assistant_echonetlite_2025-02-08T01-47-58.105Z.log)

nao-pon commented:
@SudarshanKulakrni Thanks for providing the logs! Looking at the raw data, we can see that F5 is 16 bytes and F6 is 32 bytes. However, since "\ff" continues from the 17th byte of F6 onwards, the data may not be very meaningful.

First, I created a debug quirk file that divides the first 16 bytes of both properties into 4-byte chunks, calculates them with int and signed int, and outputs 8 entities for each. Please check whether any of the 16 entities match the actual values.
First, just to be sure, update edge to the latest version, then place the following file in custom_components/echonetlite/quirks/Nichicon/all, restart HA, and observe. (file extension is "py.txt", so change it to ".py".)

- [02A5.py](https://github.com/user-attachments/files/18718735/02A5.py.txt)
```python
from homeassistant.components.sensor.const import (
    CONF_STATE_CLASS,
    SensorDeviceClass,
    SensorStateClass,
)
from homeassistant.const import (
    CONF_NAME,
    CONF_TYPE,
    CONF_UNIT_OF_MEASUREMENT,
    UnitOfEnergy,
)
from pychonet.lib.epc_functions import _int, _signed_int
from ....const import TYPE_DATA_DICT


def _02A5debug(edt):
    i1 = i2 = i3 = i4 = s1 = s2 = s3 = s4 = None
    try:
        i1 = _int(edt[0:4])
        i2 = _int(edt[4:8])
        i3 = _int(edt[8:12])
        i4 = _int(edt[12:16])

        s1 = _signed_int(edt[0:4])
        s2 = _signed_int(edt[4:8])
        s3 = _signed_int(edt[8:12])
        s4 = _signed_int(edt[12:16])
    except:
        pass
    finally:
        return {
            "i1": i1,
            "i2": i2,
            "i3": i3,
            "i4": i4,
            "s1": s1,
            "s2": s2,
            "s3": s3,
            "s4": s4,
        }


def _02A5F5(edt):
    d1 = d2 = None
    try:
        d1 = _signed_int(edt[0:4])
        d2 = _signed_int(edt[4:8])
    except:
        pass
    finally:
        return {"household_consumption": d1, "from/to_grid": d2}


def _02A5F6(edt):
    d1 = d2 = None
    try:
        d1 = float(_int(edt[0:4])) / 1000
        d2 = float(_int(edt[4:8])) / 1000
    except:
        pass
    finally:
        return {"normal_direction": d1, "reverse_direction": d2}


QUIRKS = {
    0xF5: {
        # "EPC_FUNCTION": _02A5F5,
        "EPC_FUNCTION": _02A5debug,
        "ENL_OP_CODE": {
            CONF_NAME: "Instantaneous amount of energy",
            CONF_TYPE: SensorDeviceClass.POWER,
            CONF_STATE_CLASS: SensorStateClass.MEASUREMENT,
            # TYPE_DATA_DICT: ["household_consumption", "from/to_grid"],
            TYPE_DATA_DICT: ["i1", "i2", "i3", "i4", "s1", "s2", "s3", "s4"],
        },
    },
    0xF6: {
        # "EPC_FUNCTION": _02A5F6,
        "EPC_FUNCTION": _02A5debug,
        "ENL_OP_CODE": {
            CONF_NAME: "Cumulative energy",
            CONF_TYPE: SensorDeviceClass.ENERGY,
            CONF_STATE_CLASS: SensorStateClass.TOTAL_INCREASING,
            # CONF_UNIT_OF_MEASUREMENT: UnitOfEnergy.KILO_WATT_HOUR,
            # TYPE_DATA_DICT: ["normal_direction", "reverse_direction"],
            TYPE_DATA_DICT: ["i1", "i2", "i3", "i4", "s1", "s2", "s3", "s4"],
        },
    },
}
```

SudarshanKulakrni commented:
Hi @nao-pon san,

Thanks for the update. I could see that the property "power i2" or "power s2" is able to show the consumption of the house. Thats a good step. 

I am guessing that "Power s1" is the power being bought from Grid. Here is the photo I had taken in early morning, when there was no solar production. (Sorry could not capture logs early morning, just screen capture from mobile.)

 
![image](https://github.com/user-attachments/assets/ae71299e-3288-4ccd-adda-e344dfe4e5c4)


Currently, there is a solar production happening and no power is being purchased from Grid. Here are the current parameters.

![image](https://github.com/user-attachments/assets/f3a27aec-2ca1-4313-9d19-5c9b953dccc2)
![image](https://github.com/user-attachments/assets/d4c2113e-e72b-4d34-955c-004a57666ccf)

Here is the photo of the panel  : 

![image](https://github.com/user-attachments/assets/80724b2f-dbd6-4ead-a137-3bc3c7890137)


Here is the Log while Solar panels produced electricity.

 
[home-assistant_echonetlite_2025-02-09T01-45-07.382Z.log](https://github.com/user-attachments/files/18722828/home-assistant_echonetlite_2025-02-09T01-45-07.382Z.log)

My guess : Addition of "Power i2" and "solar_measured_instantaneous_charging_discharging_electric_energy" minus "Power s1" will give total throughput of the solar panels. What do you think ?

nao-pon commented:
Yes. I'd like you to observe a little more and see if the third and fourth data of F5 have any meaning.

In edge (9d5ffcb), I added "Instantaneous power some_a" and "Instantaneous power some_b", so if you know what they are, please let me know. I will confirm the entity name. Also, in this edge, I fixed the bug that the quirks entity name was not set.

Also, since the battery is a separate object, it may be better to create the calculation of the total throughput of the solar panel using the HA sensor template.

nao-pon commented:
I found this blog post interesting, it might be worth a try.

- https://blog.hiroaki.jp/blog/2023/02/26/programmatically-set-recharge-and-discharge-times-for-nichicon-storage-batteries/

nao-pon commented:
Edge has been updated to edge(f2be74b).

SudarshanKulakrni commented:
Hi @nao-pon Thank you for late night work !
Here are the latest findings. Panel looks like below : 

![image](https://github.com/user-attachments/assets/ca976db2-0940-47dd-9093-3b3b945d0358)

Production = 2.2 Kw
Battery being charged at 1.8 Kw.

So household consumption would be 0.4 Kw.

Here are the entities : 
![image](https://github.com/user-attachments/assets/b4d08013-863a-4ba3-8cf6-88d705ef3317)

IMHO, The "Instantaneous Power from/toGrid" should be "Instantaneous power household_consumption" and Visa-versa.
("some_a" and "some_b" are still not making any sense to me ! Wonder what is being sent on these fields).

Attaching logs if these help.

Thanks again @nao-pon 
[home-assistant_echonetlite_2025-02-10T04-25-17.087Z.log](https://github.com/user-attachments/files/18728133/home-assistant_echonetlite_2025-02-10T04-25-17.087Z.log)

PS : Found good method to combine entities from multiple objects using the "Helper" (Devices and Services -> Helpers Tab).

nao-pon commented:
According to the specifications, the "Measured instantaneous amount of electricity" value is the instantaneous power value from the PCS to the grid or home. However, the specifications say that the value should be negative at this time, but the data shows a positive value, which is confusing.

![250210-144550](https://github.com/user-attachments/assets/e29f2ece-2a07-4705-87c7-ffb902978934)
> (10) Measured instantaneous electric power
This property indicates the measured effective instantaneous electric power in watts. The value range
shall be: 0x80000001-0x7FFFFFFD (-2147483647-2147483645). The “normal direction” shall be
defined as “positive”, while the “reverse direction” is defined as “negative”.

Also, I believe that the HomeSolarPower device is configured, but are the values ​​consistent with those?

SudarshanKulakrni commented:
> According to the specifications, the "Measured instantaneous amount of electricity" value is the instantaneous power value from the PCS to the grid or home. However, the specifications say that the value should be negative at this time, but the data shows a positive value, which is confusing.

Currently, when the system is operating just on battery, the value is negative. Battery discharge and "Measured instantaneous amount of electricity" show same values. 

![image](https://github.com/user-attachments/assets/593bb09f-a0ee-4450-a9d9-958f67618c22)

>Also, I believe that the HomeSolarPower device is configured, but are the values ​​consistent with those?

What is "HomeSolarDevice" device ? Please let me know if I am missing anything.

nao-pon commented:
Are there any of the following devices configured on the same IP address? In some cases, you may not have this many properties.
![Screenshot_20250210-224247](https://github.com/user-attachments/assets/b3e4c628-3358-43af-9640-382a45483e0d)
![Screenshot_20250210-224258](https://github.com/user-attachments/assets/3e84557c-e331-47fa-ac65-696da9b3fa6e)

SudarshanKulakrni commented:
> Are there any of the following devices configured on the same IP address? In some cases, you may not have this many properties.

Hi @nao-pon  Yes, there are 3 devices configured on same IP address. 

![image](https://github.com/user-attachments/assets/a24a9997-4ae6-45f6-ab10-23b518363e89)

If I put all of them on one dashboard, they look like below : 

![image](https://github.com/user-attachments/assets/14ee464e-61f0-461d-8cff-bae4f2efb039)

nao-pon commented:
Since I don't have the actual device, it's difficult for me to guess what the data not in the specifications means. I would like to correctly define "Instantaneous Power some_a", "Instantaneous Power from/to_grid", and "Instantaneous Power household_consumption" before releasing the product.

From the screenshots so far, it seems that the following formula holds true if we allow for an error of a few watts.

> "Measured instantaneous amount of electricity" ≒ ”Instantaneous Power some_a” - "Instantaneous Power from/to_grid" + "Instantaneous Power household_consumption" 


Can you fix the meaning of these three parameters while taking into account the sensor values ​​of the solar panel, battery, etc.?

SudarshanKulakrni commented:
> From the screenshots so far, it seems that the following formula holds true if we allow for an error of a few watts.
> 
> > "Measured instantaneous amount of electricity" ≒ ”Instantaneous Power some_a” - "Instantaneous Power from/to_grid" + "Instantaneous Power household_consumption"

Thank you so much for your help @nao-pon  san.
1. Please rename the entities, the "Instantaneous Power from/toGrid" should be "Instantaneous power household_consumption" and Visa-versa. 
I think the formula should be 
"Measured instantaneous amount of electricity" = "measured_instantaneous_charging_discharging_electric_energy" + "Instantaneous Power from/to_grid" + "Instantaneous Power household_consumption"

"Instantaneous Power from/to_grid" becomes negative when we are buying electricity from grid.  So a negative value from above formula will attribute to "Buying electricity". (+/- few Watts)

here is the current status at my home.

![image](https://github.com/user-attachments/assets/f4aff98e-5840-4233-9406-6c514c844a6f)

![image](https://github.com/user-attachments/assets/7c092404-1b2e-4938-8116-1bb3f58353bd)

SudarshanKulakrni commented:
Hello @nao-pon  
I kept a watch on PCS entities for last week. I think none of the "instantaneous" entities are adding up to the total solar production from solar panels. :( 

e.g. Current situation : 
![IMG_5612](https://github.com/user-attachments/assets/da9a5133-07aa-4619-873e-b0cb730e9d59)

The PCS entities are as below : 

The "Other Origin" entity has being showing "Unknown" for the entire week I tested.

![IMG_5613](https://github.com/user-attachments/assets/dd1e9da5-4352-4b9e-8f88-f804884fcfcf)

I will try go through the specifications you have shared (I am not great with Japanese though) and see if I can find something.

Thanks for all your work !

nao-pon commented:
@SudarshanKulakrni This approach focuses on properly decoding the obtained data and determining what it means. We do not generally add or subtract values, so I think that this will be defined by the user using the HA helper. I assume that the reason why the "Other Origin" entity is not obtained is because the device is not connected, so there is no problem.

In order to seek more information, I will release it as a new version in this state. Please continue to verify it, and if there is anything that needs to be fixed, please post a new issue.

Crim commented:
Hope you don't mind me piggy backing off of this thread.

I've got the home assistant integration installed via HACS, but when I attempt to register the above Nichicon remote control by putting in its IP address I always get a `Device could not be detected` error.  

Is there something missing to get this to talk to the remote control?  I've verified that the remote control can be contacted over the network (via ping, rlogin, etc...) and that it is listening on the echonet lite port (3610).  There must be something simple I'm missing?

Appreciate the help!
